### PR TITLE
fix(protocols): require unique server_label for MCP tools

### PR DESCRIPTION
**Pull request by zhaowenzi**
### Problem

The current MCP tool validation is not aligned with the OpenAI spec:
- It allows MCP tools to omit tools[].server_label (e.g., providing only server_url), while the OpenAI spec expects a server_label to identify the MCP server/tool instance.
- It does not prevent duplicate server_label values (including case-only differences), which can create ambiguous or conflicting MCP tool identifiers within a single request.
This spec mismatch can lead to inconsistent client/server behavior and unclear allowed tool filter when multiple MCP tools are present.

### Solution

Make validation OpenAI spec compatible by requiring every MCP tool to provide a non-empty server_label, and enforcing that all MCP server_label values are unique within the tools array (case-insensitive). Provide clear, index-addressable validation errors to help callers fix payloads quickly.

## Changes

- Update MCP tool validation in protocols/src/responses.rs (OpenAI spec compatible):
  - Require non-empty tools[idx].server_label for type = "mcp"; return missing_required_parameter including the parameter path.
  - Enforce unique MCP server_label values within tools (case-insensitive); return mcp_tool_duplicate_server_label.
  - Track seen labels via a HashSet of normalized (to_lowercase()) labels.
- Update spec tests in model_gateway/tests/spec/responses.rs:
  - Rename/repurpose the MCP validation test to assert server_label is required (instead of server_url).
  - Add a test ensuring duplicate server_label values (e.g., Foo/foo) are rejected.

## Test Plan

image

<details>
<summary>Checklist</summary>

- [x] `cargo +nightly fmt` passes
- [x] `cargo clippy --all-targets --all-features -- -D warnings` passes
- [ ] (Optional) Documentation updated

</details>

## Summary by CodeRabbit

* **New Features**
  * Added public helpers for generating IDs and normalizing input items.

* **Bug Fixes**
  * MCP tools now require a non-empty server label; validation returns precise, index-aware errors.
  * Enforces case-insensitive uniqueness of MCP server labels and reports duplicates.
  * Validates server-label format (must start with a letter; only alphanumeric, dash, underscore allowed) and reports invalid formats.

* **Tests**
  * Updated and added tests covering missing, duplicate, and invalid server-label cases.

### CODE REVIEW

File: protocols/src/responses.rs [hunk 1]
```diff
@@ -1125,9 +1128,30 @@ fn validate_response_tools(tools: &[ResponseTool]) -> Result<(), validator::Vali
...
+                    .map(|s| s.trim())
+                    .filter(|s| !s.is_empty())
+                else {
+                    let mut e = validator::ValidationError::new("missing_required_parameter");
+                    e.message = Some(
+                        format!("Missing required parameter: 'tools[{}].server_label'.", idx)
+                            .into(),
+                    );
+                    return Err(e);
+                };
```
[the remark this answers is not in the document]
Comment (zhaowenzi) [reply protocols/src/responses.rs]: Not an issue

File: protocols/src/responses.rs [hunk 2]
```diff
@@ -1125,9 +1128,30 @@ fn validate_response_tools(tools: &[ResponseTool]) -> Result<(), validator::Vali
...
+                    let mut e = validator::ValidationError::new("mcp_tool_duplicate_server_label");
+                    e.message = Some(
+                        format!(
+                            "Duplicate MCP server_label '{}' found in 'tools' parameter.",
+                            label
+                        )
+                        .into(),
+                    );
                     return Err(e);
                 }
```
[the remark this answers is not in the document]
Comment (zhaowenzi) [reply protocols/src/responses.rs]: Not an issue

File: protocols/src/responses.rs [hunk 3]
```diff
@@ -1,7 +1,7 @@
 // OpenAI Responses API types
 // https://platform.openai.com/docs/api-reference/responses
 
-use std::collections::HashMap;
+use std::collections::{HashMap, HashSet};
```
[the remark this answers is not in the document]
Comment (zhaowenzi) [reply protocols/src/responses.rs:4]: Not an issue

File: protocols/src/responses.rs [hunk 4]
```diff
@@ -1115,7 +1115,10 @@ fn validate_input_item(item: &ResponseInputOutputItem) -> Result<(), validator::
 
 /// Validates ResponseTool structure based on tool type
 fn validate_response_tools(tools: &[ResponseTool]) -> Result<(), validator::ValidationError> {
-    for tool in tools {
+    // MCP server_label must be present and unique (case-insensitive).
+    let mut seen_mcp_labels: HashSet<String> = HashSet::new();
+
+    for (idx, tool) in tools.iter().enumerate() {
```
[the remark this answers is not in the document]
Comment (zhaowenzi) [reply protocols/src/responses.rs:1117]: Not an issue

File: protocols/src/responses.rs [hunk 5]
```diff
@@ -1125,9 +1128,30 @@ fn validate_response_tools(tools: &[ResponseTool]) -> Result<(), validator::Vali
...
-                if tool.server_url.is_none() && tool.server_label.is_none() {
-                    let mut e = validator::ValidationError::new("mcp_tool_missing_connection_info");
-                    e.message = Some("MCP tool must have either server_url or server_label".into());
+                let Some(label) = tool
+                    .server_label
+                    .as_ref()
+                    .map(|s| s.trim())
+                    .filter(|s| !s.is_empty())
+                else {
+                    let mut e = validator::ValidationError::new("missing_required_parameter");
```
Comment (CatherineSue) [inline protocols/src/responses.rs]: nit: can we import ValidationError at the top of the file?

Comment (zhaowenzi) [reply protocols/src/responses.rs]: HI @CatherineSue 
This file uses validator::ValidationError::new(...) in many places. Do you want me to update all of them or keep this PR scoped to the MCP changes?

Comment (slin1237) [reply protocols/src/responses.rs]: we can do that in this PR

Comment (zhaowenzi) [reply protocols/src/responses.rs]: Done

File: model_gateway/tests/spec/responses.rs [hunk 6]
```diff
@@ -653,18 +653,43 @@ fn test_validate_tools_function_missing() {
...
 #[test]
-fn test_validate_tools_mcp_missing_url() {
+fn test_validate_tools_mcp_valid_ok() {
     let request = ResponsesRequest {
         input: ResponseInput::Text("test".to_string()),
         tools: Some(vec![ResponseTool {
             r#type: ResponseToolType::Mcp,
             function: None,
-            server_url: None, // Missing server_url
+            server_url: None, // server_url is optional when server_label is provided
```
Comment (key4ng) [inline model_gateway/tests/spec/responses.rs:664]: should this also be required since we don't support `connector_id` now

> server_url: optional string
The URL for the MCP server. One of server_url or connector_id must be provided.

Comment (zhaowenzi) [reply model_gateway/tests/spec/responses.rs:664]: From OpenAI spec, yes, url is required too (we don't have connector_id), but initially, smg supports only label to use static mcp server, so I will keep this logic here
